### PR TITLE
[core] Extend Shellwords (`String.shellescape`, `Array.shelljoin`) for Windows support

### LINKED ISSUE #13213: `Shellwords.shellescape` is not cross platform safe
`shellescape` is used ~234 times in the fastlane codebase:
https://github.com/fastlane/fastlane/search?q=shellescape&type=Code

Unfortunately is seems that is not cross platform safe at all :(

https://ruby-doc.org/stdlib-2.5.1/libdoc/shellwords/rdoc/Shellwords.html#method-c-shellescape

> Escapes a string so that it can be safely used in a Bourne shell command line. str can be a non-string object that responds to to_s.
> 
> Note that a resulted string should be used unquoted and is not intended for use in double quotes nor in single quotes.

First sentence already tells us that this is for "a Bourne shell", which Windows command lines are certainly not.
Second sentence then tells us that the result should not be quoted - which is necessary on Windows for paths with spaces, not matter it escaped or not.

This is independently confirmed by the great StackOverflow:
https://stackoverflow.com/a/5703599/252627
https://stackoverflow.com/a/50436815/252627

--- 

From previous testing of fastlane on Windows I know that this _mostly_ is not a problem. Most of the strings that `shellescape` is applied on do not contain spaces, so no quotes are necessary on Windows. The rest it does, does not hurt at least.

But there are cases where 

a) there could be spaces in the strings (see https://github.com/fastlane/fastlane/issues/13212) 
or 
b) there are single quotes applied to the resulting string before usage.

https://github.com/fastlane/fastlane/blob/6a658a4cf9c56c3c838aed397166c2770009b2d2/fastlane_core/lib/fastlane_core/itunes_transporter.rb#L213-L214

---

What can we do about that?

### PR DESCRIPTION
**Pull request by janpio**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

[`Shellwords.shellescape` is not cross platform safe](https://github.com/fastlane/fastlane/issues/13213). This causes problems for commands that are to be used on the shell and contain e.g. paths with spaces or other non-[A-Z0-9] characters.

### Description

This PR monkey patches the `String.shellescape` and `Array.shelljoin` methods that `Shellwords` usually supplies. The patch uses a new module `CrossplatformShellwords` that decides if the normal `Shellwords` or a new `WindowsShellwords` should be used to call the methods.

The PR fixes all the tests that started failing on Windows after the monkey patch became active: all the tests that had hardcoded expected command strings that were compared to command strings generated by the actions (mostly via `Actions.sh` internally) via `var.shellescape`. 

The PR also adds tests for the monkey patched methods, both `WindowsShellwords` and the `*.shell*` methods (in a mock Windows and non-Windows environment) are tested.

#### Limitations

`WindowsShellwords` itself is a pretty basic implementation of mainly `shellescape`. I don't expect this version to cover all use cases and edge cases. Those will be discovered over time when fastlane actions will fail because of malformed commands that do not run on Windows as expected - but it is a starting point.

Tests for actions and methods that compare a hardcoded string to a manually built command string could **not** be identified automatically during this PR. I hope there aren't to many, but it will only affect usage on Windows anyway. macOS is not affected.

This PR also could do nothing to actually _test_ the generated commands, as during test execution all the commands are just returned from `Actions.sh` and not executed. Those tests will have to happen in actual usage of the actions.

--- 
closes #13213

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: fastlane/lib/fastlane/actions/add_git_tag.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def self.run(options)
         cmd << ["-am #{message.shellescape}"]
         cmd << '--force' if options[:force]
         cmd << '-s' if options[:sign]
-        cmd << "'#{tag}'"
+        cmd << tag.shellescape
```
Comment (janpio) [inline fastlane/lib/fastlane/actions/add_git_tag.rb]: this changes the behavior a bit as it doesn't wrap the param in `'` any more, but as the tag itself is now `shellescape`d this should not cause any problem.

File: fastlane/spec/actions_specs/add_git_tag_spec.rb [hunk 2]
```diff
@@ -14,7 +14,9 @@
           add_git_tag
         end").runner.execute(:test)
 
-        expect(result).to eq("git tag -am builds/test/1337\\ \\(fastlane\\) \'builds/test/1337\'")
+        message = "builds/test/1337 (fastlane)"
+        tag = "builds/test/1337"
+        expect(result).to eq("git tag -am #{message.shellescape} #{tag.shellescape}")
```
Comment (janpio) [inline fastlane/spec/actions_specs/add_git_tag_spec.rb]: matching the previous comment, `'` are gone here as well and replaced with `shellescape`

File: fastlane_core/spec/core_ext/shellwords_ext_spec.rb [hunk 3]
```diff
@@ -0,0 +1,372 @@
+# references
```
Comment (joshdholtz) [inline fastlane_core/spec/core_ext/shellwords_ext_spec.rb]: Should this file actually be called `shellwords_spec.rb` to match with the new `shellwords.rb`?

File: fastlane_core/lib/fastlane_core/core_ext/shellwords.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+require_relative '../helper'
+require 'shellwords'
+
+# Here be monkey patches
+
+class String
+  # CrossplatformShellwords
+  def shellescape
```
Comment (joshdholtz) [inline fastlane_core/lib/fastlane_core/core_ext/shellwords.rb]: Does this 100% override the method defined by `shellwords` because this is required after `shellwords`?

File: fastlane_core/lib/fastlane_core/core_ext/shellwords.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    CrossplatformShellwords.shellescape(self)
+  end
+end
+
+class Array
+  def shelljoin
```
Comment (joshdholtz) [inline fastlane_core/lib/fastlane_core/core_ext/shellwords.rb]: Same question here - Does this 100% override the method defined by `shellwords` because this is required after `shellwords`?

**Review by joshdholtz — COMMENTED** [review]
Few small questions! Otherwise this looks 💯

File: fastlane_core/spec/core_ext/shellwords_ext_spec.rb (hunk 3 above)
Comment (janpio) [reply fastlane_core/spec/core_ext/shellwords_ext_spec.rb]: I chose the name after the other file `\fastlane_core\spec\core_ext\cfpropertylist_ext_spec.rb` which contains the tests for an implementation in `\fastlane_core\lib\fastlane_core\core_ext\cfpropertylist.rb`. But the other one would work as well, of course.

File: fastlane_core/lib/fastlane_core/core_ext/shellwords.rb (hunk 4 above)
Comment (janpio) [reply fastlane_core/lib/fastlane_core/core_ext/shellwords.rb]: I am not sure what you are asking, but I mostly did this by how the other stuff was done - so possible I missed something.

My current understanding:
1. `fastlane_core/lib/fastlane_core.rb` requires `fastlane_core/core_ext/shellwords` which loads it from `/lib/fastlane_core/core_ext/shellwords.rb`.
2. There we require 'shellwords, which loads the native implementation.
3. Then the monkey patching happens.

Comment (janpio) [reply fastlane_core/lib/fastlane_core/core_ext/shellwords.rb]: If your question is if this overwrites the `String.shellescape` defined by `Shellwords` completely, then: Yes. It completely deletes that method and replaces it with this code - which for non-Windows environments does the exact same thing as shellword's `String.shellescape` - handing the string over to `Shellwords.escape`.

Comment (joshdholtz) [reply fastlane_core/lib/fastlane_core/core_ext/shellwords.rb]: Ah yes, your assumption of what I was asking in your last response was correct 👌 Thanks for the answer! ❤️

**Review by joshdholtz — DISMISSED** [review]
Two small requests about adding assertion that `Shellwords.escape` is actually getting called 🙃

File: fastlane_core/spec/core_ext/shellwords_ext_spec.rb [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+  end
+
+  describe "on other OSs (macOS, Linux)" do
+    before(:each) do
+      allow(FastlaneCore::Helper).to receive(:windows?).and_return(false)
+    end
+
+    os = 'other'
+    shelljoin_testcases.each do |testcase|
+      it testcase['it'] + ': ' + testcase['it_result'][os] do
```
[the remark this answers is not in the document]
Comment (janpio) [reply fastlane_core/spec/core_ext/shellwords_ext_spec.rb]: On non-Windows...

Yes, good idea. Will add this.

[the remark this answers is not in the document]
Comment (janpio) [reply fastlane_core/spec/core_ext/shellwords_ext_spec.rb]: Turned out to be a bit of work, especially as I had to learn about `.and_call_original` first 🍇 

But well, not it checks if the correct methods of the correct classes are called.

File: fastlane_core/spec/core_ext/shellwords_ext_spec.rb [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+  end
+
+  describe "on other OSs (macOS, Linux)" do
+    before(:each) do
+      allow(FastlaneCore::Helper).to receive(:windows?).and_return(false)
+    end
+
+    os = 'other'
+    shellescape_testcases.each do |testcase|
+      it testcase['it'] + ': ' + testcase['it_result'][os] do
```
[the remark this answers is not in the document]
Comment (janpio) [reply fastlane_core/spec/core_ext/shellwords_ext_spec.rb]: Done, see new commit.

**Review by joshdholtz — APPROVED** [review]
@janpio This looks good 👍 Thanks for making those testing changes and answering my questions ❤️
